dot/sync/chain_sync.go

```diff
@@ -0,0 +1,784 @@
+		return big.NewInt(2<<32 - 1)
+	}
+
+	for _, ps := range cs.peerState {
+		sum = big.NewInt(0).Add(sum, ps.number)
+		count++
+	}
+
+	return big.NewInt(0).Div(sum, big.NewInt(count))
+}
+
+// handleWork handles potential new work that may be triggered on receiving a peer's state
+// in bootstrap mode, this begins the bootstrap process
+// in tip mode, this adds the peer's state to the pendingBlocks set and potentially starts
+// a fork sync
+func (cs *chainSync) handleWork(ps *peerState) error {
+	logger.Trace("handling potential work", "target hash", ps.hash, "target number", ps.number)
+	worker, err := cs.handler.handleNewPeerState(ps)
+	if err != nil {
+		return err
+	}
+
+	if worker == nil {
+		return nil
+	}
+
+	cs.tryDispatchWorker(worker)
+	return nil
+}
+
+func (cs *chainSync) tryDispatchWorker(w *worker) {
+	// if we already have the maximum number of workers, don't dispatch another
+	if len(cs.workerState.workers) >= MAX_WORKERS {
+		logger.Trace("reached max workers, ignoring potential work")
+		return
+	}
+
+	// check current worker set for workers already working on these blocks
+	// if there are none, dispatch new worker
+	if cs.handler.hasCurrentWorker(w, cs.workerState.workers) {
+		return
+	}
+
+	cs.workerState.add(w)
+	go cs.dispatchWorker(w)
+}
+
+// dispatchWorker begins making requests to the network and attempts to receive responses up until the target
+// if it fails due to any reason, it sets the worker `err` and returns
+// this function always places the worker into the `resultCh` for result handling upon return
+func (cs *chainSync) dispatchWorker(w *worker) {
+	logger.Debug("dispatching sync worker",
+		"target hash", w.targetHash,
+		"target number", w.targetNumber,
+	)
+
+	if w.targetNumber == nil || w.startNumber == nil {
+		logger.Error("must provide a block start and target number",
+			"startNumber==nil?", w.startNumber == nil,
+			"targetNumber==nil?", w.targetNumber == nil,
+		)
+		return
+	}
+
+	start := time.Now()
+	defer func() {
+		end := time.Now()
+		w.duration = end.Sub(start)
+		logger.Debug("sync worker complete", "success?", w.err == nil)
+		cs.resultQueue <- w
+	}()
+
+	reqs, err := workerToRequests(w)
+	if err != nil {
+		// if we are creating valid workers, this should not happen
+		logger.Crit("failed to create requests from worker", "worker", w, "error", err)
+		return
+	}
+
+	for _, req := range reqs {
+		// TODO: if we find a good peer, do sync with them, right now it re-selects a peer each time
+		err := cs.doSync(req)
+		if err != nil {
+			// failed to sync, set worker error and put into result queue
+			w.err = err
+			return
+		}
+	}
+}
+
+func (cs *chainSync) doSync(req *network.BlockRequestMessage) *workerError {
+	// determine which peers have the blocks we want to request
+	peers := cs.determineSyncPeers(req)
+
+	if len(peers) == 0 {
+		cs.Lock()
+		for p := range cs.ignorePeers {
+			delete(cs.ignorePeers, p)
+		}
```

Comment:
instead of removing one at a time you could do `cs.ignorePeers = cs.ignorePeers[:0]`, afaik

Reply:
can't do this as `cs.ignorePeers` is a map, not a slice. I could just create a new empty map, but not sure which would be faster/slower

Reply:
ok, I thought it was a slice, no problems so.